### DISCUSSION TITLE
MySQL Support for DKIM Keys

### DISCUSSION THREAD

**Discussion by adamleemiller** [rspamd/rspamd#4993] [Q&A]
Hello! I am curious about the ability to pull DKIM keys from a MySQL database. The set up is Postfix, Dovecot, Rspamd, OpenDKIM, and MySQL with a custom built web application for managing domains, mailboxes and aliases. Everything is working well right now except DKIM validation does not work because it is looking for the keys locally however this will not work as we create different keys for each domain name which are then stored in the database. These keys can also be rotated by the simple click of a button within our custom user interface. I should mentioned we do have Redis for history just in case this matters.

I did try the following configuration from ChatGPT which did not work:

```
enabled = true;

sql {
  selector = "SELECT subdomain FROM dkim WHERE domainName = '$domain'";
  domain = "SELECT domainName FROM dkim WHERE domainName = '$domain'";
  key = "SELECT privateKey FROM dkim WHERE domainName = '$domain' AND subdomain = '$selector'";
}

mysql {
  servers = "";
  user = "";
  password = "";
  dbname = "";
  table = "";
}
```

Any help is greatly appreciated. 

Thank you!

**Comment by fatalbanana** [comment]
> I am curious about the ability to pull DKIM keys from a MySQL database.

There is not such a feature nor direct support for MySQL in general.

> I did try the following configuration from ChatGPT which did not work:

It doesn't know how to configure Rspamd & invents garbage.

Way to integrate this could be via some HTTP service- but it'd call for patching the plugin.

If you have custom integration with Rspamd and can supply a key in an HTTP header on scan-time that is another possibility.

**Reply by adamleemiller** [reply]
I am curious. With Postfix and Dovecot being heavily integrated with *SQL for support virtual domains / mailboxes and rspamd being written to integrate directly with Postfix and the like, why is there no support for *SQL? Not to mention, Gmail does not sign every single email it sends with the same DKIM key for every single email address on its service with a custom domain so there's definitely a need to store it somewhere other than a local file. Sure, I could put the keys in Redis but that adds more complexity to an already complex environment especially when you consider how much usage Postfix + Dovecot + *SQL already receive in the hosting world. It just seems odd to not have *SQL support.

I would be more than happy to work through a custom solution using an HTTP service, I would just need to know where to start. My custom UI has an API which I could use to expose the keys via an HTTP call, I would just need to know how to go about integrating that request into rspamd.

Thank you
